server/ServerApp.cpp

```diff
@@ -3371,6 +3372,14 @@ void ServerApp::PostCombatProcessTurns() {
     m_universe.BackPropagateObjectMeters();
     empires.BackPropagateMeters();
 
+    // add new techs
```

Comment:
Moving the adding of new techs to here, after scripted turn events and executing meters on newly-created objects means that tech research won't have any effect on new objects created during the turn processing after a player clicks turn. That could be potentially annoying or problematic in some cases, where having the tech researched would have some benefit to ships produced when the tech is researched. Similarly, any scripted turn events would execute as though the tech was not finished yet, which means one needs to finish researching techs a full turn prior to getting their benefit / effect in that context.

I'm unsure if either of these are really substantial issues though, so are not major concerns to me now...

Reply:
If I understand correctly after reading a bit more of the related and background threads, I think you intend to delay the effects from a newly-researched tech from executing by one turn.

So, the player would see the "tech researched" sitrep when receiving the updates for turn N, but the effects of that tech won't contribute to the actual meter values (and resource output or detection range) that the player sees and gets until turn N+1. Presumably the tech effects would still be apparent in the next turn predicted meter values the player sees, though.

The code difference is adding of techs to the empire after, instead of before, the `ApplyAllEffectsAndUpdateMeters` and `BackPropagateObjectMeters` calls. The back propagation sets the actual meter values that will be used for the subsequent turn processing, and the `UpdateMeterEstimates` below will just update the expected next turn meter values (which may change when the player submits orders).

This has an advantage, in that the expected next turn meter values are more accurate... researching a tech won't cause them to change between when the player submits orders and when they get back the results of the next turn (although other unknown or unpredictable factors may make the estimates inaccurate).

It does also mean that it takes an extra turn between being informed a tech is researched and getting the benefits of the tech. This is potentially annoying / disappointing for players, although a bit less so in that they should be able to see the differences in the expected next turn meter values on the same turn the player gets the tech researched sitrep.